{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68ad9bfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: biopython in c:\\users\\khush\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (1.86)\n",
      "Requirement already satisfied: numpy in c:\\users\\khush\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from biopython) (1.26.4)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "[notice] A new release of pip is available: 25.0.1 -> 25.3\n",
      "[notice] To update, run: C:\\Users\\khush\\AppData\\Local\\Microsoft\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\python.exe -m pip install --upgrade pip\n"
     ]
    }
   ],
   "source": [
    "! pip install biopython\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "20305047",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Annotation Information:\n",
      "Gene name: BRCA1\n",
      "Description: BRCA1 DNA repair associated\n",
      "Chromosome: Not available\n",
      "Protein Product: Not available\n",
      "Gene Symbol: BRCA1\n",
      "Aliases: ['IRIS', 'PSCP', 'BRCAI', 'BRCC1', 'FANCS', 'PNCA4', 'RNF53', 'BROVCA1', 'PPP1R53']\n",
      "Gene Type:Not available\n",
      "Genomic Location: Not available\n",
      "Genomic Location: {'Seq-interval': {'Seq-interval_from': '43044294', 'Seq-interval_to': '43170326', 'Seq-interval_strand': {'Na-strand': StringElement('', attributes={'value': 'minus'})}, 'Seq-interval_id': {'Seq-id': {'Seq-id_gi': '568815581'}}}}\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = \"khushidriven@gmail.com\"\n",
    "\n",
    "# Step1: Enter gene name \n",
    "gene_name=input(\"Enter the name of gene: \")\n",
    "\n",
    "# Step2: Search NCBI gene database \n",
    "handle = Entrez.esearch(db=\"gene\", term=f\"{gene_name}[Gene Name] AND Homo sapiens[Organism]\")\n",
    "record = Entrez.read(handle)\n",
    "handle.close()\n",
    "\n",
    "if record[\"IdList\"]:\n",
    "    gene_id=record[\"IdList\"][0]\n",
    "    \n",
    "    \n",
    "#Step3: Gene Details \n",
    "    handle=Entrez.efetch(db=\"gene\", id=gene_id, rettype='xml')\n",
    "    records=Entrez.read(handle)\n",
    "    handle.close()\n",
    "\n",
    "    gene_info=records[0]\n",
    "\n",
    "    print(\"Gene Annotation Information:\")\n",
    "    print(f\"Gene name: {gene_name}\")\n",
    "\n",
    "# Description of gene\n",
    "    try: \n",
    "        description = gene_info[\"Entrezgene_gene\"][\"Gene-ref\"][\"Gene-ref_desc\"]\n",
    "    except:\n",
    "        description = \"Not available\"\n",
    "    print(f\"Description: {description}\")\n",
    "\n",
    "\n",
    "# Chromosome \n",
    "    try:\n",
    "        chromosome = gene_info['Entrezgene_source']['BioSource']['SubSource'][0]['SubSource_name']\n",
    "    except:\n",
    "        chromosome = \"Not available\"\n",
    "    print(f\"Chromosome: {chromosome}\")\n",
    "\n",
    "\n",
    "# Protein product \n",
    "    try: \n",
    "        products= gene_info['Entrezgene_locus'][0]['Gene-commentary_products']\n",
    "        protein_name= products[0]['Gene-commentary_text']\n",
    "        print(f\"Protein Product: {protein_name}\")\n",
    "    except:\n",
    "        print(\"Protein Product: Not available\")\n",
    "        \n",
    "# Gene Symbol \n",
    "    try:\n",
    "        gene_symbol=gene_info['Entrezgene_gene']['Gene-ref']['Gene-ref_locus']\n",
    "    except:\n",
    "        gene_symbol=\"Not available\"\n",
    "    print(f\"Gene Symbol: {gene_symbol}\")\n",
    "    \n",
    "# Aliases\n",
    "    try:\n",
    "        aliases=gene_info['Entrezgene_gene']['Gene-ref']['Gene-ref_syn']\n",
    "    except:\n",
    "        aliases=\"Not available\"\n",
    "    print(f\"Aliases: {aliases}\")\n",
    "    \n",
    "\n",
    "# Gene type\n",
    "    try:\n",
    "        gene_type=gene_info['Entrezgene_type']['value']\n",
    "    except:\n",
    "        gene_type=\"Not available\"\n",
    "    print(f\"Gene Type:{gene_type}\")\n",
    "    \n",
    "\n",
    "# Genomic location\n",
    "    try:\n",
    "        genomic_location= gene_info['Entrezgene_locus'][0]['Gene-commentary_seqs'][0]['Seq-loc_int']\n",
    "        start= genomic_location['Seq-interval_from']\n",
    "        end= genomic_location['Seq-interval_to']\n",
    "        print(f\"Genomic location:{start}-{end}\")\n",
    "    except:\n",
    "        print(\"Genomic Location: Not available\")\n",
    "    print(f\"Genomic Location: {genomic_location}\")\n",
    " \n",
    "else:\n",
    "    print(\"Gene not found in NCBI\")\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63931f0d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
